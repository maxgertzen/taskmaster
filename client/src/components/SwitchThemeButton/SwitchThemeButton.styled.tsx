import styled from '@emotion/styled';

export const StyledSwitchThemeButton = styled('div')(({ theme }) => ({
  display: 'flex',
  alignItems: 'center',
  cursor: 'pointer',
  gap: theme.spacing(1),
}));
