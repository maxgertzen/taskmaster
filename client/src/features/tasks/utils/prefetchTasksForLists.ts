import { QueryClient } from '@tanstack/react-query';

import { QUERY_KEYS } from '@/shared/api/query-keys';
import { STALE_TIME } from '@/shared/constants/staleTime';

import { fetchTasks } from '../api';

export const prefetchTasksForLists = (
  listIds: string[],
  queryClient: QueryClient,
  token: string | null
) => {
  if (!token || !listIds.length) return;

  listIds.forEach((listId) => {
    queryClient.prefetchQuery({
      queryKey: QUERY_KEYS.tasks({ listId }),
      queryFn: fetchTasks(token),
      staleTime: STALE_TIME,
    });
  });
};
