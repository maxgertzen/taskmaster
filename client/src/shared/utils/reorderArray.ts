export const reorderArray = <T>(
  array: T[],
  fromIndex: number,
  toIndex: number
): T[] => {
  const reordered = [...array];
  const [movedItem] = reordered.splice(fromIndex, 1);
  reordered.splice(toIndex, 0, movedItem);
  return reordered;
};
