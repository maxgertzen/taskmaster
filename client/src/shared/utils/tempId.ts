export const generateTempId = (prefix: string = 'temp') =>
  `${prefix}-${Date.now()}`;
