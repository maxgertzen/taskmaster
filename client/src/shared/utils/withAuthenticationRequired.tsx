import { withAuthenticationRequired as realWithAuthenticationRequired } from '@auth0/auth0-react';

import { IS_AUTH0_DISABLED } from '@/constants';
import { mockWithAuthenticationRequired } from '@/mocks';

const isMock = IS_AUTH0_DISABLED === 'true';

export const withAuthenticationRequired = isMock
  ? mockWithAuthenticationRequired
  : realWithAuthenticationRequired;
