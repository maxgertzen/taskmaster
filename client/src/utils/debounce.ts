export function debounce<T extends (...args: Parameters<T>) => ReturnType<T>>(
  func: T,
  delay: number
): ((...args: Parameters<T>) => void) & { cancel: () => void } {
  let timeoutId: ReturnType<typeof setTimeout> | null = null;

  const debouncedFunction = (...args: Parameters<T>) => {
    if (timeoutId) clearTimeout(timeoutId);
    timeoutId = setTimeout(() => func(...args), delay);
  };

  debouncedFunction.cancel = () => {
    if (timeoutId) clearTimeout(timeoutId);
  };

  return debouncedFunction;
}
