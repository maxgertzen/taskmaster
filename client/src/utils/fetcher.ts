export type FetcherConfig<K> = {
  url: string;
  method?: 'GET' | 'POST' | 'PUT' | 'DELETE';
  body?: K;
  timeout?: number;
  headers?: HeadersInit;
  urlSearchParams?: string;
};

export const fetcher = async <T, K = undefined>(
  config: FetcherConfig<K>
): Promise<T> => {
  try {
    const {
      method = 'GET',
      body = {},
      headers = {},
      timeout = 5000,
      urlSearchParams,
    } = config;
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), timeout);

    const options: RequestInit = {
      method,
      headers: {
        'Content-Type': 'application/json',
        ...headers,
      },
      ...(method !== 'GET' && { body: JSON.stringify(body) }),
      signal: controller.signal,
    };

    if (urlSearchParams) {
      config.url += `?${urlSearchParams}`;
    }

    const response = await fetch(config.url, options);
    clearTimeout(timeoutId);

    if (!response.ok) {
      throw new Error(`Error: ${response.status} ${response.statusText}`);
    }

    return await response.json();
  } catch (error) {
    console.error('Failed to fetch data:', error);
    throw error;
  }
};
