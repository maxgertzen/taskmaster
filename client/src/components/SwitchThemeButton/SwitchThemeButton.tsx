import { FC } from 'react';

import { useThemeStore } from '../../store/themeStore';
import { SpriteIcon } from '../SpriteIcon/SpriteIcon';

import { StyledSwitchThemeButton } from './SwitchThemeButton.styled';

interface SwitchThemeButtonProps {
  withText?: boolean;
}
export const SwitchThemeButton: FC<SwitchThemeButtonProps> = ({ withText }) => {
  const activeTheme = useThemeStore((state) => state.theme);
  const toggleTheme = useThemeStore((state) => state.toggleTheme);

  const handleToggle = () => {
    toggleTheme();
  };

  if (withText) {
    return (
      <StyledSwitchThemeButton role='button' onClick={toggleTheme}>
        <SpriteIcon
          alt='switch theme'
          name={activeTheme === 'light' ? 'moon' : 'sun'}
        />
        <label>{activeTheme === 'light' ? 'Dark' : 'Light'}</label>
      </StyledSwitchThemeButton>
    );
  }

  return (
    <SpriteIcon
      alt='switch theme'
      name={activeTheme === 'light' ? 'moon' : 'sun'}
      onClick={handleToggle}
    />
  );
};
