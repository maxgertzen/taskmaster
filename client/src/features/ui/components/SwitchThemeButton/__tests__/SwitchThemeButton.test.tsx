import { fireEvent } from '@testing-library/dom';
import { act } from '@testing-library/react';
import { describe, it, expect, vi, beforeEach } from 'vitest';

import {
  mockToggleTheme,
  mockThemeStoreMode,
  mockThemeStore,
} from '@/tests/__mocks__/storeMocks';
import { render, screen } from '@/tests/utils';

import { SwitchThemeButton } from '../SwitchThemeButton';

vi.mock('@/shared/store/themeStore', () => ({
  useThemeStore: mockThemeStore,
}));

describe('SwitchThemeButton', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  describe.each([
    ['light', 'moon', 'Dark'],
    ['dark', 'sun', 'Light'],
  ])('in %s mode', (theme, expectedIcon, expectedLabel) => {
    beforeEach(() => {
      mockThemeStoreMode.mockReturnValue(theme);
    });

    it('renders correctly without text', () => {
      const { container } = render(<SwitchThemeButton />);
      expect(container).toMatchSnapshot();
      expect(screen.getByTestId(`icon-${expectedIcon}`)).toBeInTheDocument();
    });

    it('renders correctly with text', () => {
      const { container } = render(<SwitchThemeButton withText />);
      expect(container).toMatchSnapshot();
      expect(screen.getByText(expectedLabel)).toBeInTheDocument();
    });

    it('calls toggleTheme when clicked (without text)', async () => {
      render(<SwitchThemeButton />);

      await act(async () => {
        fireEvent.click(screen.getByTestId(`icon-${expectedIcon}`));
      });
      expect(mockToggleTheme).toHaveBeenCalled();
    });

    it('calls toggleTheme when clicked (with text)', async () => {
      render(<SwitchThemeButton withText />);

      await act(async () => {
        fireEvent.click(screen.getByTestId('switch-theme-button'));
      });
      expect(mockToggleTheme).toHaveBeenCalled();
    });
  });
});
