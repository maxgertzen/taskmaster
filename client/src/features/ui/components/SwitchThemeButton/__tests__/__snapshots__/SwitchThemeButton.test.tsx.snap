// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SwitchThemeButton > in dark mode > renders correctly with text 1`] = `
<div>
  <div
    class="css-1mzk1eu"
    data-testid="switch-theme-button"
    role="button"
  >
    <div
      class="css-4dcg29"
      data-testid="icon-sun"
      role="button"
      size="3"
    >
      <svg
        data-testid="sun-icon"
      />
    </div>
    <label>
      Light
    </label>
  </div>
</div>
`;

exports[`SwitchThemeButton > in dark mode > renders correctly without text 1`] = `
<div>
  <div
    class="css-4dcg29"
    data-testid="icon-sun"
    role="button"
    size="3"
  >
    <svg
      data-testid="sun-icon"
    />
  </div>
</div>
`;

exports[`SwitchThemeButton > in light mode > renders correctly with text 1`] = `
<div>
  <div
    class="css-1mzk1eu"
    data-testid="switch-theme-button"
    role="button"
  >
    <div
      class="css-4dcg29"
      data-testid="icon-moon"
      role="button"
      size="3"
    >
      <svg
        data-testid="moon-icon"
      />
    </div>
    <label>
      Dark
    </label>
  </div>
</div>
`;

exports[`SwitchThemeButton > in light mode > renders correctly without text 1`] = `
<div>
  <div
    class="css-4dcg29"
    data-testid="icon-moon"
    role="button"
    size="3"
  >
    <svg
      data-testid="moon-icon"
    />
  </div>
</div>
`;
